keps/sig-storage/1979-object-storage-support/README.md

```diff
@@ -3,703 +3,742 @@
+    |      "parameters": {                     |
+    |         "key": "value"                   |
+    |      }                                   |
+    | })                                       |
+    |------------------------------------------|
+```
+Once the Bucket is successfully created in the OSP, sidecar sets `BucketReady`condition to `True`. At this point, the Bucket is ready to be utilized by workloads.
 
-![Object Relationships](images/object-rel.png)
+## Generating Access Credentials for Buckets
 
-# Workflows
-Here we describe the workflows used to automate provisioning of new and existing buckets, and the de-provisioning of these buckets.
+The following stakeholders are involved in the lifecycle of access credential generation:
 
-When the app pod is started, the kubelet will gRPC call _NodeStageVolume_ and _NodePublishVolume_, which are received by the cosi-node-adapter. The pod waits until the adapter responds to the gRPC requests. The adapter ensures that the target bucket has been provisioned and is ready to be accessed.
+ - Users  - request access to buckets
+ - Admins - establish cluster wide access policies
 
-When the app pod terminates, the kubelet gRPC calls _NodeUnstageVolume_ and _NodeUnpublishVolume_, which the cosi-node-adapter also receives. The adapter orchestrates tear-down of the ephemeral volume provisioned for this bucket.
+Access credentials are represented by BucketAccess objects. Here are the steps for creating a BucketAccess:
 
-General prep:
-+ admin creates the needed bucket classes and bucket access classes.
+  ![COSI Access Credential Creation](images/create-bucket-access.jpg)
 
-Prep for brownfield:
-+ admin creates `Bucket` instances representing backend buckets.
+###### 1. Admin creates BucketAccessClass, User requests BucketAccess to be created
 
-## Create Bucket
+The BucketAccessClass represents a set of common properties shared by multiple BucketAccesses. It is used to specify policies for creating access credentials, and also for configuring driver-specific access parameters. More information about BucketAccessClass is [here](#bucketaccessclass)
 
-![CreateBucket Workflow](images/create-bucket.png)
+The BucketAccessRequest is a request to create BucketAccess. This resource is designed to specify the Bucket for which the credentials provide access, and also for specifying access modes (rw, ro, wo) for that the application requires. More information about BucketAccessRequest is [here](#bucketaccessrequest)
 
-This workflow describes the automation supporting creating a new (greenfield) backend bucket. Accessing this bucket is covered in [Sharing COSI Created Buckets](#sharing-cosi-created-buckets). COSI determines that a request is for an existing bucket when `BucketRequest.bucketInstanceName` is specified.
+BucketSecret is the name of the secret that COSI will generate with credentials to access the bucket. The same secret name has to be set in the podSpec as a projected secret volume.
```

Comment:
In the case where the secret is rotated we will have a new set of files, what would the user need to do? Would the user need application code to acknowledge the fact that there are new secrets? (NOTE: I read Bucket Provisioning (for pods) I understand that the user needs to have code to watch the `bucket.json` file)

Reply:
After much discussion, we abandoned the concept of secret rotation. Mainly because no mechanism exists today to ensure the client that uses this credentials, can be forced to use the rotated access/secret keys. 

> (NOTE: I read Bucket Provisioning (for pods) I understand that the user needs to have code to watch the bucket.json file)

The user just needs to read it upfront. There is no need to watch the file. As of the current design, there is no plan to dynamically update the file

Reply:
I believe that using long term keys is a security concern, the cloud providers have created mechanism like https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp.html to avoid doing this

I also found the following in https://docs.aws.amazon.com/general/latest/gr/aws-access-keys-best-practices.html:

> Rotate access keys periodically. Change access keys on a regular basis. For details, see Rotating Access Keys (AWS CLI, Tools for Windows PowerShell, and AWS API) in the IAM User Guide and How to Rotate Access Keys for IAM Users on the AWS Security Blog.

And in https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html

> As a best practice, use temporary security credentials (IAM roles) instead of access keys, and disable any AWS account root user access keys. For more information, see Best Practices for Managing AWS Access Keys in the Amazon Web Services General Reference.

Reply:
I just read your comment about this feature not being supported in non-cloud storage providers, I would make an argument in favor of the direction the industry is going (which seems to be the usage of an Authorization server) and ask which are the non cloud OSPs that require access tokens to work? also what would their path moving forward be? are they considering having their own auth service too?

Reply:
As an example Google Cloud has a Workload Identity server that can run in GKE and also in a Bare Metal/Multi Cloud env, ref https://cloud.google.com/anthos/multicluster-management/fleets/workload-identity#outside

Reply:
There is no standard mechanism for doing serviceAccount based authentication without forcing applications to change their credential discovery pattern. Ideally, we want to go to a metadata style service, where the application gets the latest token from a link-local source. Just like instance IAM in AWS, or the equivalent in GCP. 

We punted this to post-alpha because there is a lot more groundwork needed for service account based auth, specifically the part where we tie Service Accounts to OSP authentication. This would be step 1.

Secondly, we cannot force applications to start using COSI style credential discovery upfront. Our opinion is, this would also be DoA if we don't have the simpler, more crude but workable approach of access/secret credentials. Then, as the next step,  we will add the more secure pattern, and allow applications to opt-in. We have already started discussions on adding COSI as an auth provider for major cloud platform SDKs, so that applications can seamlessly start using COSI. This would be step 2.

We want to make the transition from no COSI -> COSI an easy one to promote adoption. As of alpha, I think credentials is a good start.

Reply:
Copying from slack FTR: I'm not saying every implementation has to support SA tokens, I am saying that the API has to allow for implementations that do.

Reply:
> we cannot force applications to start using COSI style credential discovery upfront

In the context of access tokens, I believe that by introducing `bucket.json` instead of the mechanisms the cloud providers set you're forcing the applications to change their credential discovery approach upfront anyways. If I have an app that's expecting the existence of `~/.aws/credentials` or the env vars `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` wouldn't it be better to use the [Secret projection as a file](https://kubernetes.io/docs/concepts/configuration/secret/#projection-of-secret-keys-to-specific-paths) or the [Secret projection of env vars](https://kubernetes.io/docs/concepts/configuration/secret/#using-secrets-as-environment-variables) so that I don't have to write additional code in my app to read the `bucket.json` file? I think that even though the intent is to normalize all the types of credentials into a single file it'd be great to list the pros/cons of not projecting them using k8s mechanisms.

Reply:
yeah, we were already planning on adding serviceAccount based auth post-alpha. It needs further discussion, but it is definitely on the roadmap.